lib/statsd/instrument/batched_udp_sink.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+module StatsD
+  module Instrument
+    # @note This class is part of the new Client implementation that is intended
+    #   to become the new default in the next major release of this library.
+    class BatchedUDPSink
+      DEFAULT_FLUSH_INTERVAL = 1.0
+      MAX_PACKET_SIZE = 508
+
+      def self.for_addr(addr, flush_interval: DEFAULT_FLUSH_INTERVAL)
+        host, port_as_string = addr.split(":", 2)
+        new(host, Integer(port_as_string), flush_interval: flush_interval)
+      end
+
+      attr_reader :host, :port
+
+      def initialize(host, port, flush_interval: DEFAULT_FLUSH_INTERVAL)
+        @host = host
+        @port = port
+        @mutex = Mutex.new
+        @socket = nil
+        @flush_interval = flush_interval
+
+        require "concurrent/array"
+        @buffer = Concurrent::Array.new
+        @dispatcher_thread = nil
+        spawn_dispatcher
+      end
+
+      def sample?(sample_rate)
+        sample_rate == 1 || rand < sample_rate
+      end
+
+      def <<(datagram)
+        @buffer << datagram
+        spawn_dispatcher unless @dispatcher_thread&.alive?
+        self
+      end
+
+      def shutdown
```

Comment:
I added this method for test purpose. However if for some reason an app is creating lots of short lived sinks, I'm pretty sure the dispatcher thread will prevent them from being garbage collected, causing a memory leak and worse and thread leak.